compiler/src/iree/compiler/Dialect/Flow/Transforms/SinkReshapes.cpp

```diff
@@ -64,35 +78,79 @@ static bool shouldSinkExpandShapeOp(OpOperand *opOperand) {
   // cloned into their producers.
   if (isDequantizationLikeOp(consumer)) {
     return false;
   }
 
-  // If the op is already fusable with producer using tile and fuse,
-  // do nothing.
-  if (llvm::any_of(consumer->getOpOperands(), [](OpOperand &opOperand) {
-        Operation *currProducer = opOperand.get().getDefiningOp();
-        Operation *currConsumer = opOperand.getOwner();
-        return isFusableUsingTileAndFuse(currProducer, currConsumer) &&
-               // The check for the producer having a single use is not fully
-               // worked out. Ideally we can fuse with a producer irrespective
-               // of number of uses, but is a good thumb rule in practice.
-               llvm::hasSingleElement(currProducer->getUses());
-      })) {
+  // First check that the expand_shape producer and consumer can be fused.
+  Operation *reshapeProducer = reshapeOp.getSrc().getDefiningOp();
+  if (!reshapeProducer) {
     return false;
   }
-
-  // Do not sink if consumer is a contraction/matmul like op.
-  if (auto linalgConsumerOp = dyn_cast<linalg::LinalgOp>(consumer)) {
-    if (linalg::isaContractionOpInterface(linalgConsumerOp))
-      return false;
+  if (!isFusableUsingTileAndFuse(reshapeOp.getSrc().getDefiningOp(),
+                                 consumer)) {
+    return false;
   }
 
-  return isFusableUsingTileAndFuse(reshapeOp.getSrc().getDefiningOp(),
-                                   consumer);
+  // If the op is already fusable with producer using tile and fuse,
+  // do nothing.
+  for (OpOperand &opOperand : consumer->getOpOperands()) {
+    Operation *currProducer = opOperand.get().getDefiningOp();
+    if (!currProducer) {
+      continue;
+    }
+
+    // The check for the producer having a single use is not fully
+    // worked out. Ideally we can fuse with a producer irrespective
+    // of number of uses, but is a good thumb rule in practice.
+    if (!llvm::hasSingleElement(currProducer->getUses())) {
+      continue;
+    }
+
+    // Check if a producer can already be tiled and fused with the consumer.
+    if (!isFusableUsingTileAndFuse(currProducer, consumer)) {
+      continue;
+    }
+
+    // There is already a tile-and-fusable producer to fuse with. Still prefer
+    // fusing with the producer whose parallel iteration space rank matches
+    // the consumer parallel iteration space rank to avoid loss of parallelism.
+    if (isa<linalg::LinalgOp>(reshapeProducer) &&
+        isa<linalg::LinalgOp>(currProducer)) {
+      unsigned currConsumerNumParallelLoops =
+          consumerGenericOp.getNumParallelLoops();
+      unsigned currProducerNumParallelLoops =
+          cast<linalg::LinalgOp>(currProducer).getNumParallelLoops();
+      if (currProducerNumParallelLoops == currConsumerNumParallelLoops) {
+        // If the producer has same number of parallel loops as consumer,
+        // then this is the operand to fuse along. So do nothing.
+        return false;
+      }
```

Comment:
I think the logic here is faulty. If the reshape producer is not a linalg op, then it will skip this whole block and skip the `currProducer` even if it had equivalent parallel loops. Don't we still want to return `false` if the `currProducer` has equivalent parallel loops but the `reshapeProducer` is not a linalg op?

Reply:
Line 102 implicitly checks that `reshapeProducer` is a linalg op. The `isa<linalg::LinalgOp>(reshapeProducer)` is redundant. Maybe should be moved to line 102

Also, it should probably be changed to: `!isFusableUsingTileAndFuse(reshapeProducer, consumer)` for clarity

Reply:
I think it can still be an unpack or unset_encoding op though, in which case this may cause the reshape propagation to force fusion with the `reshapeProducer` unpack/unset_encoding op instead of `currProducer` when the `currProducer` has an equivalent number of parallel loops.

Reply:
Thanks for looking into the logic. More eyes on it is great!

So unset encoding actually does not have iteration spaces (unpack does though..). Let me try to account for this better.